src/main/java/spoon/support/reflect/declaration/CtModuleImpl.java

```diff
@@ -0,0 +1,467 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.support.reflect.declaration;
+
+import spoon.SpoonException;
+import spoon.reflect.annotations.MetamodelPropertyField;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtModule;
+import spoon.reflect.declaration.CtModuleDirective;
+import spoon.reflect.declaration.CtPackage;
+import spoon.reflect.declaration.CtPackageExport;
+import spoon.reflect.declaration.CtProvidedService;
+import spoon.reflect.declaration.CtModuleRequirement;
+import spoon.reflect.declaration.CtUsedService;
+import spoon.reflect.factory.ModuleFactory;
+import spoon.reflect.path.CtRole;
+import spoon.reflect.reference.CtModuleReference;
+import spoon.reflect.visitor.CtVisitor;
+import spoon.support.comparator.CtLineElementComparator;
+import spoon.support.util.SortedList;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class CtModuleImpl extends CtNamedElementImpl implements CtModule {
+	@MetamodelPropertyField(role = CtRole.MODIFIER)
+	private boolean openModule;
+
+	@MetamodelPropertyField(role = CtRole.MODULE_DIRECTIVE)
+	private List<CtModuleDirective> moduleDirectives = CtElementImpl.emptyList();
+
+	@MetamodelPropertyField(role = CtRole.SUB_PACKAGE)
+	private CtPackage rootPackage;
+
+	public CtModuleImpl() {
+		super();
+	}
+
+	@Override
+	public boolean isUnnamedModule() {
+		return TOP_LEVEL_MODULE_NAME.equals(this.getSimpleName());
+	}
+
+	@Override
+	public boolean isOpenModule() {
+		return this.openModule;
+	}
+
+	@Override
+	public <T extends CtModule> T setModuleDirectives(List<CtModuleDirective> moduleDirectives) {
+		getFactory().getEnvironment().getModelChangeListener().onListDeleteAll(this, CtRole.MODULE_DIRECTIVE, this.moduleDirectives, new ArrayList<>(this.moduleDirectives));
+		if (moduleDirectives == null || moduleDirectives.isEmpty()) {
+			this.moduleDirectives = CtElementImpl.emptyList();
+			return (T) this;
+		}
+		if (this.moduleDirectives == CtElementImpl.<CtModuleDirective>emptyList()) {
+			this.moduleDirectives = new SortedList<>(new CtLineElementComparator());
```

Comment:
I suggest to sort directives only once at time when CtModule is created by JDT builder. I think, later automatic sorting of client specific module directives by position is not expected by clients. WDYT?

Reply:
I don't agree. I think we should document that the module directives are ordered by position. Now I think users don't put position attribute in the element when they're modifying the model. So they will use the method `addModuleDirectiveAtPosition` when they want to specify a position.